src/app/pipes/meql.pipe.spec.ts

```diff
@@ -0,0 +1,356 @@
+limitations under the License.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+import { MeqlPipe } from './meql.pipe';
+import { TestBed } from '@angular/core/testing';
+import moment from 'moment';
+
+const newline = '\r\n';
+const tab = '\t';
+
+class meqlTestHelper {
+  public static singleConditionTest(
+    pipe: MeqlPipe,
+    field: string,
+    operator: string,
+    value: any
+  ) {
+    /*
+  Expected example:
+
+  (
+    "Field" = "Value"
+  )
+   */
+    const query = {
+      condition: 'and',
+      rules: [{ field, operator, value }],
+    };
+    const line1 = `(${newline}`;
+    const line2 = `${tab}"${query.rules[0].field}" ${query.rules[0].operator} "${query.rules[0].value}"${newline}`;
+    const line3 = ')';
+    return [pipe.transform(query), `${line1}${line2}${line3}`];
+  }
+
+  public static doubleConditionTest(pipe: MeqlPipe, query: any) {
+    /*
+  Expected example:
+
+  (
+    "Field" = "Value"
+    and "Field2" = "Value2"
+  )
+   */
+    const line1 = `(${newline}`;
+    const line2 = `${tab}"${query.rules[0].field}" ${query.rules[0].operator} "${query.rules[0].value}"${newline}`;
+    const line3 = `${tab}${query.condition} "${query.rules[1].field}" ${query.rules[1].operator} "${query.rules[1].value}"${newline}`;
+    const line4 = ')';
+    return [pipe.transform(query), `${line1}${line2}${line3}${line4}`];
+  }
+
+  public static doubleRulesetTest(pipe: MeqlPipe, query: any) {
+    /*
+  Expected example:
+
+  (
+    "Field" = "Value"
+    and "Field2" = "Value2"
+    and (
+        "Field3" = "Value3"
+        or "Field4" = "Value4"
+    )
+  )
+   */
+    const line1 = `(${newline}`;
+    const line2 = `${tab}"${query.rules[0].field}" ${query.rules[0].operator} "${query.rules[0].value}"${newline}`;
+    const line3 = `${tab}${query.condition} "${query.rules[1].field}" ${query.rules[1].operator} "${query.rules[1].value}"${newline}`;
+    const line4 = `${tab}${query.condition} (${newline}`;
+    const line5 = `${tab}${tab}"${query.rules[2].rules[0].field}" ${query.rules[2].rules[0].operator} "${query.rules[2].rules[0].value}"${newline}`;
+    const line6 = `${tab}${tab}${query.rules[2].condition} "${query.rules[2].rules[1].field}" ${query.rules[2].rules[1].operator} "${query.rules[2].rules[1].value}"${newline}`;
+    const line7 = `${tab})${newline}`;
+    const line8 = ')';
+    return [
+      pipe.transform(query),
+      `${line1}${line2}${line3}${line4}${line5}${line6}${line7}${line8}`,
+    ];
+  }
+}
+
+describe('JqlPipe', () => {
+  let pipe: MeqlPipe;
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      providers: [MeqlPipe],
+    });
+    pipe = TestBed.inject(MeqlPipe);
+  });
+
+  it('create an instance', () => {
+    expect(pipe).toBeTruthy();
+  });
+
+  it('transforms field equal to string', () => {
+    const testResult = meqlTestHelper.singleConditionTest(
+      pipe,
+      'Simple String Field',
+      '=',
+      'Some Text'
+    );
+    expect(testResult[0]).toBe(testResult[1]);
```

Comment:
No problem here, but there is another way to do this which is hopefully more self-explanatory. Looking at your test helper, the function returns an array with just 2 objects in it, so it is effectively a tuple. You can destructure tuples in JS/TS into individual variables, so you could write this test like so:

```ts
it('transforms field equal to string', () => {
    const [actual, expected] = meqlTestHelper.singleConditionTest(
      pipe,
      'Simple String Field',
      '=',
      'Some Text'
    );
    expect(actual).toBe(expected);
  });
```

The reason I suggest it is because referring to `actual` and `expected` seems clearer than understanding what `testResult[0]` and `testResult[1]` is.

Reply:
I agree this would be clearer. Updated as suggested.

Reply:
...yet superseded by further refactoring so no longer relevant.